### ISSUE TITLE
Bug in processing escaped chars in response_template

### ISSUE THREAD

**Issue by ChanderG** [foundation-model-stack/fms-hf-tuning#492]
## Describe the bug

One would use a response_template of the form "\n### Response:" normally. The first "\n" is included to encode the tokens correctly with context (see [here] (https://huggingface.co/docs/trl/en/sft_trainer#using-tokenids-directly-for-responsetemplate) for some explanation). In the [current code](https://github.com/foundation-model-stack/fms-hf-tuning/blob/c3aa25ccb2048b8d7c09acd30f85c19f83e28db8/tuning/data/data_preprocessing_utils.py#L67), the prefix tokens are stripped.

Some models like Llama have wierd tokenizer behaviour where the text tokenizes differently with a newline. For example:

```
>>> from transformers import AutoTokenizer
>>> tok_l = AutoTokenizer.from_pretrained("meta-llama/Llama-3.1-8B")
>>> msg = "###Input: This is an input. \n###Response:\n This is the output."
>>> tok_l(msg)
{'input_ids': [128000, 14711, 2566, 25, 1115, 374, 459, 1988, 13, 720, 14711, 2647, 512, 1115, 374, 279, 2612, 13], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
>>> tok_l("\n###Response:")
{'input_ids': [128000, 198, 14711, 2647, 25], 'attention_mask': [1, 1, 1, 1, 1]}
>>> tok_l("\n###Response:\n")
{'input_ids': [128000, 198, 14711, 2647, 512], 'attention_mask': [1, 1, 1, 1, 1]}
```

This means, we need to use response_template with an extra "\n" at the end. In my testing, this was not a problem with Granite, but this is definitely needed with Llama.

But, using this new response template didn't solve the problem. Instead a new error showed up:
```
UserWarning: Could not find response key `[2647,  7338, 77]` in the following instance...
```

This was puzzling, since the expected response key is `[2647, 512]`. Decoding this new token set shows that the input text recieved is "Response:\\n". This is especially puzzling since just printing out the response_template at all points shows the single "\n", but in reality, internally the string stored is actually "\\n". Somehow this problem has happened only for the trailing "\n", perhaps because the first "\n" is anyway thrown away after tokenizing.

TL;DR: When specifying a response_template of the form "###Response:\n", internally the string being stored is actually "###Response:\\n". Thus, after tokenizing, the tokens are not found in the input data due to this mismatch and leads to the usual "Could not find response key.... this input will not be used in the loss calculations.".

The solution for this is quite simple and a PR is included below.

## Platform

v2.5.0

## Sample Code

Included above.

## Expected behavior

Correct tokenizing.

## Observed behavior

Included above.

## Additional context

The fix is straightforward and included in a linked PR.

**Comment by kmehant** [comment]
Just adding to this discussion, this is actually a serious problem, which token comes right after/before the response template can eventually modify the way response template is tokenized making it unfindable in the actual token stream. This can be true to any such markings (instruction template etc) used for masking. Unfortunately today, completiononly collator (https://github.com/huggingface/trl/blob/e3244d2d096ff1e2e248c931d06d39e165e20623/trl/trainer/utils.py#L109-L115) does only work on token ids instead of string match to split and mask as an option. We may need to think on options to solve this more resiliently such that given marking is a string subset of the stream then it should surely find and mask. Few options could be

1. Strictly enforce use of special tokens as the only option so that response template will always be tokenized to the same token id irrespective of this neighbour text.
2. Extend collators (mainly completiononly collator) to support text based matching rather just token id based match and mask.

**Comment by dushyantbehl** [comment]
Closed as we fixed few things in the code related to this, can reopen if needed later.
